parsil/src/assembler.rs

```diff
@@ -371,12 +396,18 @@ impl<'a, C: ContextProvider> Assembler<'a, C> {
                         };
                         let bound = Some(self.expression_to_boundary(right)?);
 
-                        if self.secondary_index_bounds.low.is_some() {
-                            // impossible to say which is higher between two
-                            // conflicting low bounds
-                            self.secondary_index_bounds.low = None;
+                        if bounds.low.is_some() {
+                            match bound.partial_cmp(&bounds.low) {
+                                // If $sid > x, $sid > y and y > x, then $sid > y
+                                Some(Ordering::Greater) => bounds.low = bound,
+                                // If $sid > x, $sid > y and x > y, then $sid > x
+                                Some(_) => {}
+                                // impossible to say which is higher between two
+                                // conflicting low bounds
+                                None => bounds.low = None,
```

Comment:
Can't we still leave one of them in `bounds`, even if they are potentially contradicting each other? Even if we have 2 bounds which might be conflicting, we can use one to filter rows to be proven, and evaluate the other predicate on all the filtered rows? Like if we have predicates `sec_index > $1 + 2` and `sec_index > $2`, we can use one of them as the query bound, say `sec_index > $2`; in this way, we provide as proving rows only the rows where `sec_index > $2`, but on each of these rows the circuit will still evaluate the predicate `sec_index > $1 + 2`, so at the end the query result should be correct. This reasoning should allow to have quite always query bounds that could significantly cut the number of rows to be proven for a query. Wdyt?

Reply:
No, that wouldn't be possible because the way we compute the secondary index bounds is included in the query hash for security, so we cannot change it at execution time depending on the values of the placeholders. 

What I meant is that, even if we can determine which bound is more restrictive at compile-time, keeping one of them should still be better than not having any bounds. Indeed, if we keep one bound, we will restrict the set of rows that need to be proven, and so we will evaluate the other predicate only on this set of rows, and this shouldn't change the results of the query I think. Does it make sense or am I missing some issues?

Reply:
I see. So the `AND` in the circuits is short-cutting, meaning that if we have two potential bounds `AND`-ed, then we don't need to iterate over the range coming from the `OR` of the two bounds to prove that we checked all the rows – covering from either bound is enough?

Reply:
Exactly, for completeness the circuits will only consider the bounds computed as specified in `QueryBoundSecondary` basically, so as long as we put bounds which covers a superset of the possible matching rows of a query, it should be fine